### PR TITLE
New gather subcommand to assist debugging bootstrap failures.

### PR DESCRIPTION
**Pull request by jstuever**
Adds a new subcommand to assists with debugging a failed bootstrap by collecting logs from the cluster. Currently, it outputs the shell commands to run in order to gather these logs. It can later be extended to gather the logs itself. It is called automatically from the create cluster subcommand on a failed bootstrap.

https://jira.coreos.com/browse/CORS-1050

### CODE REVIEW

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -266,6 +266,10 @@ func waitForBootstrapComplete(ctx context.Context, config *rest.Config, director
 	}, 2*time.Second, apiContext.Done())
 	err = apiContext.Err()
 	if err != nil && err != context.Canceled {
+		err2 := runGatherBootstrapCmd(nil, nil)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go]: IMO it makes sense to call this after we have decided that waitForBootstrap has failed [here](https://github.com/openshift/installer/pull/1627/files#diff-266f03122200ace2d21f2e449b6a5503R99)

File: cmd/openshift-install/create.go [hunk 2]
```diff
@@ -302,6 +306,12 @@ func waitForEvent(ctx context.Context, client cache.Getter, name string, timeout
 		},
 	)
 
+	if name == "bootstrap-complete" {
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go]: IMO same as https://github.com/openshift/installer/pull/1627/files#r275891274

File: cmd/openshift-install/create.go (hunk 1 above)
Comment (jstuever) [reply cmd/openshift-install/create.go]: Agree, done.

Comment (wking) [reply cmd/openshift-install/create.go]: > IMO it makes sense to call this after we have decided that `waitForBootstrap` has failed here

Don't we want `create cluster` failures to spit this out too?  I'd expect we want it in `waitForBootstrapComplete` where `create cluster` and `wait-for bootstrap-complete` can share it.

File: cmd/openshift-install/gather.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	logrus.Info("Use the following commands to gather logs from the cluster")
+	logrus.Infof("> ssh core@%s -A '/usr/local/bin/installer-gather.sh %s'", bootstrap, strings.Join(masters, " "))
+	logrus.Infof("> scp core@%s:~/installer-logs.tar.gz .", bootstrap)
+
+	return nil
+}
+
+func logGatherBootstrap(boostrap string, masters []string) {
+	if s, ok := os.LookupEnv("SSH_AUTH_SOCK"); !ok || s == "" {
+		logrus.Info("Make sure ssh-agent is running, env SSH_AUTH_SOCK is set to the ssh-agent's UNIX socket and your private key is added to the agent.")
```
Comment (wking) [inline cmd/openshift-install/gather.go]: How portable is `ssh-agent` and `SSH_AUTH_SOCK`?  My impression is that this will be POSIX- (or maybe even Linux-)specific.

File: cmd/openshift-install/gather.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+		masters   []string
+	}
+)
+
+func newGatherBootstrapCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "bootstrap",
+		Short: "Gather debugging data for a failing to bootstrap control plane",
+		Args:  cobra.ExactArgs(0),
+		RunE:  runGatherBootstrapCmd,
```
Comment (wking) [inline cmd/openshift-install/gather.go]: When we use `setupFileHook` to log to the file (like here, but not for `return cmd.Help()` above), `Run` with `logrus.Fatal` is better than `RunE` (it gets the final error message into the log files).  More in #786.

File: cmd/openshift-install/gather.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+	cmd.PersistentFlags().StringVar(&gatherBootstrapOpts.bootstrap, "bootstrap", "", "")
+	cmd.PersistentFlags().StringArrayVar(&gatherBootstrapOpts.masters, "master", []string{}, "")
+	return cmd
+}
+
+func runGatherBootstrapCmd(_ *cobra.Command, _ []string) error {
+	cleanup := setupFileHook(rootOpts.dir)
+	defer cleanup()
+
+	tfStateFilePath := filepath.Join(rootOpts.dir, terraform.StateFileName)
```
Comment (wking) [inline cmd/openshift-install/gather.go]: I'd much rather have the bootstrap IP stored somewhere more isolated from Terraform.  @abhinavdahiya had previously discussed storing it in `metadata.json` (although it's been a while; my memory is fuzzy).  But we could have `Cluster.Generate` add a per-platform switch so each platform could provide the bootstrap IP, and we could just drop that in a separate `created.json`.

File: cmd/openshift-install/gather.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+	}
+	//TODO: print the ssh command
+}
+
+func extractHostAddresses(config *types.InstallConfig, tfstate terraformState) (bootstrap string, masters []string, err error) {
+	mcount := *config.ControlPlane.Replicas
+	switch config.Platform.Name() {
+	case awstypes.Name:
+		bm := tfstate.Modules["root/bootstrap"]
+		bootstrap, _, err = unstructured.NestedString(bm.Resources["aws_instance.bootstrap"], "primary", "attributes", "public_ip")
```
Comment (wking) [inline cmd/openshift-install/gather.go]: Here's an example of something that would go in the AWS-specific extraction logic.  Other platforms will need different implementations.

File: cmd/openshift-install/gather.go (hunk 3 above)
Comment (abhinavdahiya) [reply cmd/openshift-install/gather.go]: the goal is to make sure the user has agent running, this is an easy check to report user that it might not.

`ssh-agent` works on both linux and macOS using the same env. and we don't do windows yet ;)

File: cmd/openshift-install/gather.go (hunk 5 above)
Comment (abhinavdahiya) [reply cmd/openshift-install/gather.go]: > @abhinavdahiya had previously discussed storing it in metadata.json (although it's been a while; my memory is fuzzy). But we could have Cluster.Generate add a per-platform switch so each platform could provide the bootstrap IP, and we could just drop that in a separate created.json, or some such.

do not want to expand the files created by installer anymore for now. the `tfstate` has everything we would need for now. This approach works fine for now as this allows us to get bootstrap ip and control plane ips for each platform.

File: cmd/openshift-install/create.go (hunk 1 above)
Comment (abhinavdahiya) [reply cmd/openshift-install/create.go]: the create cluster and wait-for cannot share this anyways because create cluster we are sure that we wouldn't need the `--{bootstrap,control-plane}-address` flags but, wait-for if might need them.

So if it was upto me i would not call this at all and just print something like `call gather bootstrap-failure subcommand` to gather logs

Comment (jstuever) [reply cmd/openshift-install/create.go]: I think this makes sense if/when this command can actually gather the logs. Else we are asking them to run one command just to output more commands for them to run. May as well get to the point for now.

File: cmd/openshift-install/waitfor.go [hunk 7]
```diff
@@ -46,6 +46,9 @@ func newWaitForBootstrapCompleteCmd() *cobra.Command {
 
 			err = waitForBootstrapComplete(ctx, config, rootOpts.dir)
 			if err != nil {
+				logrus.Info("Use the following commands to gather logs from the cluster")
+				logrus.Info("> ssh core@<bootstrap> -A '/usr/local/bin/installer-gather.sh <control-plane1> <control-plane2> <control-plane3> ...'")
+				logrus.Info("> scp core@<bootstrap>:~/log-bundle.tar.gz .")
```
Comment (sferich888) [inline cmd/openshift-install/waitfor.go]: Are these <system> lines filled in for the user? 
If not, can they be? 
If not, they can't be (scope creep), we need to document how to pull these out of some installer file. (may need to denote this as a todo somewhere).

File: cmd/openshift-install/create.go [hunk 8]
```diff
@@ -96,6 +96,10 @@ var (
 
 				err = waitForBootstrapComplete(ctx, config, rootOpts.dir)
 				if err != nil {
+					err2 := runGatherBootstrapCmd(nil, nil)
+					if err2 != nil && err2 != context.Canceled {
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go]: why the `context.Canceled`

File: cmd/openshift-install/waitfor.go [hunk 9]
```diff
@@ -46,6 +46,9 @@ func newWaitForBootstrapCompleteCmd() *cobra.Command {
 
 			err = waitForBootstrapComplete(ctx, config, rootOpts.dir)
 			if err != nil {
+				logrus.Info("Use the following commands to gather logs from the cluster")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/waitfor.go]: this should point users to run the `openshift-install --dir <root-dir> gather bootstrap` command.
and not the `ssh scp commands.`

File: cmd/openshift-install/gather.go (hunk 5 above)
Comment (wking) [reply cmd/openshift-install/gather.go]: > This approach works fine for now as this allows us to get bootstrap ip and control plane ips for each platform.

But it requires all of our platforms to use Terraform?  Anyhow, maybe ok for the short term, but I'd at least like to push the per-platform code down into per-platform subpackages so we can delegate ownership.

File: cmd/openshift-install/waitfor.go (hunk 7 above)
Comment (wking) [reply cmd/openshift-install/waitfor.go]: > If not, they can't be (scope creep), we need to document how to pull these out of some installer file. (may need to denote this as a todo somewhere).

For `create cluster` runs, we already pull it out of installer files (although see [here][1] with quibbles about how we do that).  But `wait-for bootstrap-complete` can also be run in a UPI setting, and then the installer has no way to determine the bootstrap machine's IP address, it's just using a user-provided kubeconfig to talk to the Kubernetes API, and there may be load balancers, etc. in between.

[1]: https://github.com/openshift/installer/pull/1627#discussion_r275936520

File: cmd/openshift-install/gather.go (hunk 5 above)
Comment (wking) [reply cmd/openshift-install/gather.go]: Another motivation for extracting this to Terraform (and maybe like @sferich888's question [here][1]), is "what about folks who want the bootstrap IP for other purposes?  For example, when we run the gather code from CI, it would be nice to be able to drop the Terraform -> bootstrap IP extraction and lean on this new installer code instead.

Comment (abhinavdahiya) [reply cmd/openshift-install/gather.go]: > But it requires all of our platforms to use Terraform? Anyhow, maybe ok for the short term,

we have bigger problems as a lot of other places already assume terraform. so we solve that when we get to that bridge.

> but I'd at least like to push the per-platform code down into per-platform subpackages so we can delegate ownership.

personally this is so small and i don't see this growing that i'm fine keeping it here for now.

Comment (wking) [reply cmd/openshift-install/gather.go]: > we have bigger problems as a lot of other places already assume terraform. so we solve that when we get to that bridge.

Works for me.

File: cmd/openshift-install/waitfor.go (hunk 7 above)
Comment (sferich888) [reply cmd/openshift-install/waitfor.go]: I think we need to document this clearly but it makes since. tl;dr in IPI situation we try and auto pillot the collection. In UPI, we show you how to fly, plug in your vectors.

File: cmd/openshift-install/gather.go (hunk 5 above)
Comment (sferich888) [reply cmd/openshift-install/gather.go]: What about failures where tfstate is not crated?

Comment (abhinavdahiya) [reply cmd/openshift-install/gather.go]: > What about failures where tfstate is not crated?

This is gather on bootstrap failure so I'm not sure why you think tfstate will not exist in ipi

And on UPI where we know tfstate will never exist, we require user provide both bootstrap host and control plane host addresses

File: cmd/openshift-install/gather.go (hunk 4 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: @wking modified... ptal.

File: cmd/openshift-install/create.go (hunk 8 above)
Comment (jstuever) [reply cmd/openshift-install/create.go]: @abhinavdahiya I forgot to delete it from a copy/paste. Modified, ptal.

File: cmd/openshift-install/waitfor.go (hunk 9 above)
Comment (jstuever) [reply cmd/openshift-install/waitfor.go]: Modified, ptal

File: cmd/openshift-install/gather.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	bootstrap, masters, err := extractHostAddresses(config.Config, tfstate)
+	if err != nil {
+		if err2, ok := err.(errUnSupportedGatherPlatform); ok {
+			logrus.Error(err2)
+			return unSupportedPlatformGather()
+		}
+		return errors.Wrapf(err, "failed to get bootstrap and control plane host addresses from %q", tfStateFilePath)
+	}
+
+	logrus.Info("Use the following commands to gather logs from the cluster")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: can we move this to https://github.com/openshift/installer/pull/1627/files#diff-be4790dcfa83ca14e7942b4c3b50c5f9R118 ??

File: cmd/openshift-install/gather.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+	logrus.Infof("> ssh core@%s -A '/usr/local/bin/installer-gather.sh %s'", bootstrap, strings.Join(masters, " "))
+	logrus.Infof("> scp core@%s:~/log-bundle.tar.gz .", bootstrap)
```
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: I would expect that we copy the bundle to `rootOpts.dir` like all other files for installer?

File: cmd/openshift-install/gather.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+	if err != nil {
+		if err2, ok := err.(errUnSupportedGatherPlatform); ok {
+			logrus.Error(err2)
+			return unSupportedPlatformGather()
+		}
+		return errors.Wrapf(err, "failed to get bootstrap and control plane host addresses from %q", tfStateFilePath)
+	}
+
+	logrus.Info("Use the following commands to gather logs from the cluster")
+	logrus.Infof("> ssh core@%s -A '/usr/local/bin/installer-gather.sh %s'", bootstrap, strings.Join(masters, " "))
```
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: `-A` makes more sense before `core@`

File: cmd/openshift-install/gather.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+	logrus.Infof("> scp core@%s:~/log-bundle.tar.gz .", bootstrap)
+
+	return nil
+}
+
+func logGatherBootstrap(boostrap string, masters []string) {
+	if s, ok := os.LookupEnv("SSH_AUTH_SOCK"); !ok || s == "" {
+		logrus.Info("Make sure ssh-agent is running, env SSH_AUTH_SOCK is set to the ssh-agent's UNIX socket and your private key is added to the agent.")
+	}
+	//TODO: print the ssh command
```
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: missing TODO

File: cmd/openshift-install/gather.go (hunk 12 above)
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: nit can we remove the `>`

File: cmd/openshift-install/waitfor.go [hunk 14]
```diff
@@ -46,6 +46,8 @@ func newWaitForBootstrapCompleteCmd() *cobra.Command {
...
+				logrus.Info("> openshift-install --dir <root-dir> gather bootstrap")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/waitfor.go]: nit: drop the `>`

and `openshift-install gather bootstrap --help` so that people see the flags that can be used.

File: cmd/openshift-install/gather.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+		Run: func(_ *cobra.Command, _ []string) {
+			cleanup := setupFileHook(rootOpts.dir)
+			defer cleanup()
+			err := runGatherBootstrapCmd(rootOpts.dir)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+		},
+	}
+	cmd.PersistentFlags().StringVar(&gatherBootstrapOpts.bootstrap, "bootstrap", "", "")
```
Comment (abhinavdahiya) [inline cmd/openshift-install/gather.go]: can you include description for the flags.

File: cmd/openshift-install/gather.go (hunk 11 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: I would leave it to the end user to decide where they want to copy the file, if not pwd.

File: cmd/openshift-install/gather.go (hunk 15 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: Done.

File: cmd/openshift-install/waitfor.go (hunk 14 above)
Comment (jstuever) [reply cmd/openshift-install/waitfor.go]: Done.

File: cmd/openshift-install/gather.go (hunk 12 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: Done

File: cmd/openshift-install/gather.go (hunk 13 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: Added shell commands here.

File: cmd/openshift-install/gather.go (hunk 12 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: Done.

File: cmd/openshift-install/gather.go (hunk 10 above)
Comment (jstuever) [reply cmd/openshift-install/gather.go]: Done.

**Comment by abhinavdahiya** [comment]
LGTM
ping @wking do you mind taking another look
can you squash the commits (i'm okay with just one) after ^^ and then we can move ahead with this.

File: cmd/openshift-install/waitfor.go (hunk 7 above)
Comment (jstuever) [reply cmd/openshift-install/waitfor.go]: Docs have a card to document this.

**Comment by jstuever** [comment]
@abhinavdahiya Squashed, just waiting on review labels now.

**Comment by abhinavdahiya** [comment]
> LGTM
> ping @wking do you mind taking another look
> can you squash the commits (i'm okay with just one) after ^^ and then we can move ahead with this.
> @abhinavdahiya Squashed, just waiting on review labels now.

/lgtm

we can fix nits from @wking as we iterate on this.

File: data/data/bootstrap/files/usr/local/bin/installer-gather.sh [hunk 16]
```diff
@@ -1,7 +1,6 @@
 #!/usr/bin/env bash
-set -eo pipefail
```
Comment (wking) [inline data/data/bootstrap/files/usr/local/bin/installer-gather.sh]: This completely removes error checking, no?  What happens if, for example `mkdir -p "${ARTIFACTS}/bootstrap/journals"` fails and we continue to launch all the commands that attempt to fill it in?

Comment (abhinavdahiya) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh]: I think we should try to make sure this script never fails and gathers as much as it can.

Comment (abhinavdahiya) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh]: also the bootstrap node is very controlled env so things like mkdir failing, we'll probably catch in our CI ;)

Comment (wking) [reply data/data/bootstrap/files/usr/local/bin/installer-gather.sh]: > also the bootstrap node is very controlled env so things like mkdir failing, we'll probably catch in our CI ;)

*If* we look and notice that things are missing.  I'm unlikely to be that observant unless quite a lot is missing ;).

File: cmd/openshift-install/waitfor.go (hunk 9 above)
Comment (wking) [reply cmd/openshift-install/waitfor.go]: Modified again to drop `--help` and set `--dir` in #1800.
